Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,381 @@
+{
+    while (!range.collapsed()) {
+        auto& node = range.startContainer();
+        auto offset = range.startOffset();
+        
+        // This check is not in the spec.
+        // I believe there is an error in the spec which I have filed an issue for
+        // https://github.com/WICG/scroll-to-text-fragment/issues/189
+        if (offset == node.length()) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node)) {
+                range.start = { *newStart, 0 };
+                continue;
+            }
+            return;
+        }
+        
+        if (isNonSearchableSubtree(node)) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
+                range.start = { *newStart, 0 };
+            else
+                return;
+            continue;
+        }
+
+        if (!isVisibleTextNode(node)) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
+                range.start = { *newStart, 0 };
+            else
+                return;
+            continue;
+        }
+
+        auto string = node.textContent();
+
+        if (string.substringSharingImpl(offset, 6) == "&nbsp;"_s)
+            offset += 6;
+
+        if (string.substringSharingImpl(offset, 5) == "&nbsp"_s)
+            offset += 5;
+        
+        if (!isSpaceOrNewline(string[offset]))
+            return;
+        offset++;
+        
+        if (offset >= node.length()) {
+            if (auto newStart = nextInComposedTreeIgnoringUserAgentShadow(node))
+                range.start = { *newStart, 0 };
+            else
+                return;
+        } else
+            range.start = { node, offset };
+    }
+}
+
+// https://wicg.github.io/scroll-to-text-fragment/#find-a-range-from-a-text-directive
+std::optional<SimpleRange> rangeForFragment(const ParsedTextDirective parsedTextDirective, Document& document)
+{
+    auto searchRange = makeRangeSelectingNodeContents(document);
+    std::optional<SimpleRange> matchRange;
+    WordBounded mustEndAtWordBoundary = (!parsedTextDirective.textEnd.isNull() || parsedTextDirective.suffix.isNull()) ? WordBounded::Yes : WordBounded::No;
+    std::optional<SimpleRange> potentialMatch;
+    
+    while (!searchRange.collapsed()) {
+        if (!parsedTextDirective.prefix.isNull()) {
+            auto prefixMatch = rangeOfStringInRange(parsedTextDirective.prefix, searchRange, WordBounded::Yes, WordBounded::No);
+            if (!prefixMatch)
+                return std::nullopt;
+            
+            auto newStart = makeBoundaryPoint(makeContainerOffsetPosition(prefixMatch->start).next());
+            if (!newStart)
+                return std::nullopt;
+            
+            searchRange.start = newStart.value();
+            matchRange = makeSimpleRange(prefixMatch->end, searchRange.end);
+            advanceRangeStartToNextNonWhitespace(matchRange.value());
```

Comment:
Here, we should assert that matchRange contains a SimpleRange.
Also, I'd prefer making advanceRangeStartToNextNonWhitespace returning a new Range instead of mutating the argument.